Plugins/ExaTrkX/include/Acts/Plugins/ExaTrkX/detail/ConnectedComponents.cuh

```diff
@@ -24,6 +27,42 @@ __device__ void swap(T &a, T &b) {
 
 /// Implementation of the FastSV algorithm as shown in
 /// https://arxiv.org/abs/1910.05971
+
+/// Hooking step of the FastSV algorithm
+template <typename TEdge, typename TLabel>
+__device__ void hookEdgesImpl(std::size_t i, const TEdge *sourceEdges,
+                              const TEdge *targetEdges, const TLabel *labels,
+                              TLabel *labelsNext, bool &changed) {
+  auto u = sourceEdges[i];
+  auto v = targetEdges[i];
+
+  if (labels[u] == labels[labels[u]] && labels[v] < labels[u]) {
+    labelsNext[labels[u]] = labels[v];
+    changed = true;
+    //printf("Edge (%i,%i): set labelsNext[%i] = labels[%i] = %i\n", u, v, labels[u], v, labels[v]);
+  } else if (labels[v] == labels[labels[v]] && labels[u] < labels[v]) {
+    labelsNext[labels[v]] = labels[u];
+    changed = true;
+    //printf("Edge (%i,%i): set labelsNext[%i] = labels[%i] = %i\n", u, v, labels[v], u, labels[u]);
+  } else {
+    //printf("Edge (%i,%i): no action\n", u, v);
+  }
+}
+
+/// Shortcutting step of the FastSV algorithm
+template <typename TEdge, typename TLabel>
+__device__ void shortcutImpl(std::size_t i, const TEdge *sourceEdges,
+                             const TEdge *targetEdges, const TLabel *labels,
+                             TLabel *labelsNext, bool &changed) {
+  if (labels[i] != labels[labels[i]]) {
+    labelsNext[i] = labels[labels[i]];
+    //printf("Vertex %i: labelsNext[%i] = labels[%i] = %i\n", i, i, labels[i], labels[labels[i]]);
+    changed = true;
```

Comment:
This could be a return value instead of an output parameter reference, for clarity.

Reply:
Same for the previous function.

Reply:
Hmm in cases such as this one, with more than one modifiable parameter, I actually like the use of output parameters, since it makes it a bit more consistent... But I don't have a strong opinion on this

Reply:
Fair enough!